### PR TITLE
fix port when fall back to http

### LINKED ISSUE #43618: Default Uri is wrong when dotnet publish docker image to insecure registry
### Describe the bug
When publish dotnet app to an insecure docker registry, It throw an exception. Seems the port is wrong.

registry uri should be: `http://dockerhub.shengguanda.com:80`
but the exception say it request: `http://dockerhub.shengguanda.com:443`

### To Reproduce
1. create a web/blazor app
2. add a insecure registry in csproj
```xml
    <PropertyGroup>
        <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
        <ContainerRegistry>dockerhub.shengguanda.com</ContainerRegistry>
        <ContainerRepository>sgd/test-app</ContainerRepository>
        <ContainerImageTags>$(Version)</ContainerImageTags>
        <ContainerUser>root</ContainerUser>
    </PropertyGroup>
```
4.  execute : `dotnet publish -p:PublishProfile=DefaultContainer --os linux --arch x64 -c Release -p:Version=0.0.0.1`
5. then you will see the error.
6. I have tried 8.0.401 and 9.0.100-rc.1, same result.

### Exceptions (if any)
```
BlazorApp1 failed with 1 error(s) (12.0s) → bin\Release\net8.0\linux-x64\publish\
    C:\Program Files\dotnet\sdk\9.0.100-rc.1.24452.12\Containers\build\Microsoft.NET.Build.Containers.targets(241,5): error :
      Containerize: error CONTAINER004: CONTAINER1014: Manifest pull failed.
      URI: http://dockerhub.shengguanda.com:443/v2/sgd/test-app/blobs/sha256:fe7649cc1e2bc620e86b0a0fce8c281f1c6283e9e4498f4935fd916737c67c97
      HTTP status code: BadGateway
         at Microsoft.NET.Build.Containers.DefaultBlobOperations.LogAndThrowContainerHttpException[T](HttpResponseMessage response, CancellationToken cancellationToken) in /_/src/Containers/Microsoft.NET.Build.Containers/Re      gistry/DefaultBlobOperations.cs:line 79
         at Microsoft.NET.Build.Containers.DefaultBlobOperations.ExistsAsync(String repositoryName, String digest, CancellationToken cancellationToken) in /_/src/Containers/Microsoft.NET.Build.Containers/Registry/DefaultBlo      bOperations.cs:line 40
         at Microsoft.NET.Build.Containers.Registry.<>c__DisplayClass43_0.<<PushAsync>b__0>d.MoveNext() in /_/src/Containers/Microsoft.NET.Build.Containers/Registry/Registry.cs:line 444
      --- End of stack trace from previous location ---
         at Microsoft.NET.Build.Containers.Registry.PushAsync(BuiltImage builtImage, SourceImageReference source, DestinationImageReference destination, Boolean pushTags, CancellationToken cancellationToken) in /_/src/Conta      iners/Microsoft.NET.Build.Containers/Registry/Registry.cs:line 472
         at Microsoft.NET.Build.Containers.Tasks.CreateNewImage.PushToRemoteRegistryAsync(BuiltImage builtImage, SourceImageReference sourceImageReference, DestinationImageReference destinationImageReference, Telemetry tele      metry, CancellationToken cancellationToken) in /_/src/Containers/Microsoft.NET.Build.Containers/Tasks/CreateNewImage.cs:line 246

```

### Further technical details
- Include the output of `dotnet --info`
```
.NET SDK:
 Version:           9.0.100-rc.1.24452.12
 Commit:            81a714c6d3
 Workload version:  9.0.100-manifests.7c9f0ee8
 MSBuild version:   17.12.0-preview-24422-09+d17ec720d

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\9.0.100-rc.1.24452.12\

.NET workloads installed:
Configured to use loose manifests when installing new manifests.
 [maccatalyst]
   Installation Source: VS 17.11.35303.130, VS 17.12.35209.166
   Manifest Version:    17.5.9231-net9-p7/9.0.100-preview.7
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100-preview.7\microsoft.net.sdk.maccatalyst\17.5.9231-net9-p7\WorkloadManifest.json
   Install Type:              Msi

 [maui-windows]
   Installation Source: VS 17.11.35303.130, VS 17.12.35209.166
   Manifest Version:    9.0.0-preview.7.24407.4/9.0.100-preview.7
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100-preview.7\microsoft.net.sdk.maui\9.0.0-preview.7.24407.4\WorkloadManifest.json
   Install Type:              Msi

 [aspire]
   Installation Source: VS 17.11.35303.130
   Manifest Version:    8.1.0/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.aspire\8.1.0\WorkloadManifest.json
   Install Type:        FileBased

 [android]
   Installation Source: VS 17.11.35303.130, VS 17.12.35209.166
   Manifest Version:    35.0.0-preview.7.41/9.0.100-preview.7
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100-preview.7\microsoft.net.sdk.android\35.0.0-preview.7.41\WorkloadManifest.json
   Install Type:              Msi

 [ios]
   Installation Source: VS 17.11.35303.130, VS 17.12.35209.166
   Manifest Version:    17.5.9231-net9-p7/9.0.100-preview.7
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\9.0.100-preview.7\microsoft.net.sdk.ios\17.5.9231-net9-p7\WorkloadManifest.json
   Install Type:              Msi


Host:
  Version:      9.0.0-rc.1.24431.7
  Architecture: x64
  Commit:       static

.NET SDKs installed:
  6.0.133 [C:\Program Files\dotnet\sdk]
  7.0.317 [C:\Program Files\dotnet\sdk]
  8.0.203 [C:\Program Files\dotnet\sdk]
  8.0.206 [C:\Program Files\dotnet\sdk]
  8.0.400 [C:\Program Files\dotnet\sdk]
  8.0.401 [C:\Program Files\dotnet\sdk]
  9.0.100-preview.7.24407.12 [C:\Program Files\dotnet\sdk]
  9.0.100-rc.1.24452.12 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.6 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0-preview.7.24406.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 9.0.0-rc.1.24452.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0-preview.7.24405.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 9.0.0-rc.1.24431.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.33 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.20 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.6 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.8 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.0-preview.7.24405.2 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 9.0.0-rc.1.24452.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version
Rider 2024.2.5

### PR DESCRIPTION
**Pull request by dameng324**
When fallback to http，the port should fallback to 80, if don't the port will be 443.

Already tested on my local machine.

Fixed: https://github.com/dotnet/sdk/issues/43618 , https://github.com/dotnet/sdk-container-builds/issues/588

### CODE REVIEW

File: src/Containers/Microsoft.NET.Build.Containers/FallbackToHttpMessageHandler.cs
Comment (martincostello) [inline src/Containers/Microsoft.NET.Build.Containers/FallbackToHttpMessageHandler.cs:83]: This will explicitly set the port to 80 (i.e. `http://whatever:80`) - is that what's intended, or in that scenario would you want the implied default port (i.e. `http://whatever`)? If the latter, then you can set it to `-1`.

Comment (dameng324) [reply src/Containers/Microsoft.NET.Build.Containers/FallbackToHttpMessageHandler.cs:83]: Thanks for your review. 

Maybe I should change the port only when the port is 443, as the user's registry may specified the port.

Comment (dameng324) [reply src/Containers/Microsoft.NET.Build.Containers/FallbackToHttpMessageHandler.cs:83]: I tested, If user specified a http port. the publish command will result an error:
```
Uploading layer 'sha256:28d4f52c4a876bde6af159c1e53d159d94edfab99bfdc6eda8b53be82ee29eb1' to 'dockerhub.shengguanda.com:80'
C:\Program Files\dotnet\sdk\8.0.401\Containers\build\Microsoft.NET.Build.Containers.targets(242,5): error CONTAINER1013: Failed to push to the output registry: The SSL connection could not be established, see inner excepti
on. 
```

So there is no way to specify a http port. It is another issue.

Comment (baronfel) [reply src/Containers/Microsoft.NET.Build.Containers/FallbackToHttpMessageHandler.cs:83]: I think that we should defer to the user's registry configuration here:

* if the registry url had a port, then we should use it
* if the registry url did not have a port, then we should set `-1` here, because per [the docs](https://learn.microsoft.com/en-us/dotnet/api/system.uribuilder.port?view=net-8.0#system-uribuilder-port)

> If the [Port](https://learn.microsoft.com/en-us/dotnet/api/system.uri.port?view=net-8.0#system-uri-port) property is set to a value of -1, this indicates that the default port value for the protocol scheme will be used to connect to the host.

So the 'default' value of the `http` protocol would be used in this case.

**Comment by dameng324** [comment]
```cs
        if(uriBuilder.Port == 443) // default https port, change to http default port
            uriBuilder.Port = 80;
```

if `uri.Port` is not the default https port, we should not change it. 
if it is 443, we set it to the default http port 80, The hard code 80 should be fine the , as the port 443 is hard code too.

if the code is 
```cs
        if(uriBuilder.Port == 443) // default https port, change to http default port
            uriBuilder.Port = -1;
```
It looks wired. 443 is related to 80, not -1;

**Review by baronfel — CHANGES_REQUESTED** [review]
@dameng324 please change to use the approach described [here](https://github.com/dotnet/sdk/pull/43620#discussion_r1771604943) - explicitly setting to port 80 may not 'look weird' to you, but it is not a sufficient solution for all of the use cases identified.

**Comment by dameng324** [comment]
@baronfel 
Is this ok?
```cs
if(uriBuilder.Port == 443) // default https port, change to http default port
    uriBuilder.Port = -1;
```

If ok, I will commit it.

**Comment by dameng324** [comment]
I checked the `FallbackToHttpMessageHandler._port` and `uriBuilder.Port`, they are all 443, not -1;
So I can not check whether does user set the port or not simply, as user may set 443 port for http insecure registry.

The only way is introducing a `registryName` parameter for `FallbackToHttpMessageHandler`. It may too complex for this simple issue. So maybe check `port == 443` is enough.

The change has been committed. Thanks for your patients, Please check again.

**Comment by dameng324** [comment]
BTW, I'm glad to test the package before releasing to avoid this bug.

I add insecure registry support in https://github.com/dotnet/sdk/pull/39840, but saddly it's broken when 8.0.40x released. To avoid this happenning again, I can test this feature before dotnet team releasing it if needed.

**Comment by baronfel** [comment]
@dameng324 I think we have enough information already in the `FallbackToHttpMessageHandler` to solve this correctly.

When the original registry name (possibly including a port) is converted to a Uri, the `Uri` contains flags that let us know if the Uri's port is the 'default port' or not.

```fsharp
> let reg = "localhost:5000";;
val reg: string = "localhost:5000"

> let regUri = Uri $"https://{reg}";;
val regUri: Uri = https://localhost:5000/

> regUri.Port, regUri.IsDefaultPort;;
val it: int * bool = (5000, false)
```

Similarly, if the registry data doesn't contain port information, we get that data as well:

```fsharp
> let reg = "registry.mycorp.com";;
val reg: string = "registry.mycorp.com"

> let regUri = Uri $"https://{reg}";;
val regUri: Uri = https://registry.mycorp.com/

> regUri.Port, regUri.IsDefaultPort;;
val it: int * bool = (443, true)
```

This data persists across usage of the UriBuilder:

```fsharp
> let reg = "registry.mycorp.com";;
val reg: string = "registry.mycorp.com"

> let regUri = Uri $"https://{reg}";;
val regUri: Uri = https://registry.mycorp.com/

> regUri.Port, regUri.IsDefaultPort;;
val it: int * bool = (443, true)

> let builder = UriBuilder regUri;;
val builder: UriBuilder = https://registry.mycorp.com:443/

> builder.Scheme <- "http";;
val it: unit = ()

> builder.Uri.Port, builder.Uri.IsDefaultPort;;
val it: int * bool = (443, false)

> builder.Port <- -1;;
val it: unit = ()

> builder.Uri.Port, builder.Uri.IsDefaultPort;;
val it: int * bool = (80, true)
```

Note that in this last example, setting the Port to the `-1` value changed the Port and IsDefaultPort properties on the `Uri` being constructed by the builder.

Therefore, I think in this Fallback message handler, we can accurately detect if a Port was explicitly set by the user's registry domain and only 'reset' the port for the new `http` scheme if the Port was not explicitly set.

**Review by JanKrivanek — APPROVED** [review]

**Review by baronfel — APPROVED** [review]

**Comment by baronfel** [comment]
@dotnet/product-construction had to rerun this PR due to the sdk-unified-build leg, I was told to ping you as this happens.

**Comment by dameng324** [comment]
@baronfel  I cannot open your comment on my commit https://github.com/dotnet/sdk/pull/43620/commits/3fc207f18bf01c61cb497f135238a8eef048eef5. So I reply here:

For some reason, I tested `uriBuilder.Uri.IsDefaultPort`, It is false in this case. Maybe the bug is caused by this.

**Comment by tmds** [comment]
> we can accurately detect if a Port was explicitly set by the user's registry domain and only 'reset' the port for the new http scheme if the Port was not explicitly set.

I don't think `Uri` supports seeing the difference between `https://host:443` and `https://host` which is what we would need here to be able to fall back to `http://host:443` for the first, and `http://host` for the latter.

> I changed uriBuilder.Port==443 to request.RequestUri.IsDefaultPort and tested it. It works well.

These are functionally the same.

**Comment by baronfel** [comment]
@tmds this is a great and very depressing point.

**Comment by tmds** [comment]
Actually, `Uri.OriginalString` will return the original string!

**Comment by tmds** [comment]
> Actually, Uri.OriginalString will return the original string!

We also have the `registryName` nearby when we create the `FallbackToHttpMessageHandler`.

**Comment by baronfel** [comment]
I'm a little unclear on where we are with this PR - @dameng324 can you summarize quickly what you see any remaining issues being (if any?)

**Comment by dameng324** [comment]
@baronfel this issue has been fixed except user set a 443 port insecure registry.

The solution provided by @tmds to fix the exception is good,but I am on vacation these days. If you need me to do this work, I can start tomorrow.

**Comment by baronfel** [comment]
@dameng324 no rush, enjoy your vacation! We're locking down the SDK for 9.0.1xx quite soon, but this fix could be backported to the December servicing release as soon as it's merged, no problem at all.

**Comment by dameng324** [comment]
@baronfel Hello, I rebase my code, and create a new PR(https://github.com/dotnet/sdk/pull/44050) for this issue, would you have a look?

**Pull request closed without merging**
